### PR TITLE
RULE_LAUNCH_COMPILE and RULE_LAUNCH_LINK system for nvcc_wrapper

### PR DESCRIPTION
**Pull request by jrmadsen**
- This enables Kokkos_ENABLE_CUDA=ON with a CMAKE_CXX_COMPILER to be set to a non-clang C++ compiler instead of nvcc_wrapper
- This provides a way for Kokkos + CUDA to be compiled without forcing CMAKE_CXX_COMPILER to be nvcc_wrapper or clang

### Proposed 

- `kokkos_compilation` macro
  - This will enable setting the launch rule globally and on projects, directories, targets, and source files
  - By default, `find_package(Kokkos)` will invoke `kokkos_compilation(GLOBAL)`
- `find_package(Kokkos COMPONENTS separable_compilation)`
  - This will disable `kokkos_compilation(GLOBAL)` and the user will be responsible for doing:
    - `kokkos_compilation(PROJECT)`
    - `kokkos_compilation(DIRECTORY <DIR>)`
    - `kokkos_compilation(TARGET <TARGETS...>)`
    - `kokkos_compilation(SOURCE <SOURCES...>)`

### CODE REVIEW

**Comment by jrmadsen** [comment]
@crtrott @jjwilke This is good to go as long unless the CI fails. It works based my various compilation tests locally.

@dalg24 we need to add not setting CXX to nvcc_wrapper somewhere in the Jenkins testing I think.

**Comment by jrmadsen** [comment]
Oh and in all the CUDA tests, it _will_ still be using the `kokkos_launch_compiler` command even if `CXX=nvcc_wrapper`.

**Comment by jrmadsen** [comment]
@dalg24 This should now provide this exact same behavior if `CMAKE_CXX_COMPILER=nvcc_wrapper`. Thus, all the current changes should have zero effect on any of the testing. 

Before this is merged, Jenkins needs to have new instances of all tests with `-DKokkos_ENABLE_CUDA=ON` that do not set the compiler to `nvcc_wrapper`.

**Comment by masterleinad** [comment]
There still is
```
clang-8: error: unsupported option '--diag_suppress=esa_on_defaulted_function_ignored'
```

**Comment by jrmadsen** [comment]
@masterleinad I really need to finished that PR for CDash... if only @dalg24 wasn't so damn picky... I have no idea where that error is in the log. Did that happen in the build tree or in the install tree?

**Comment by masterleinad** [comment]
It's in the build step for `CUDA-9.2-Clang` (https://cloud.cees.ornl.gov/jenkins-ci/blue/organizations/jenkins/Kokkos/detail/Kokkos/2152/pipeline).

**Comment by jrmadsen** [comment]
Thanks, I think just needed logic for Clang

File: cmake/KokkosConfig.cmake.in
Comment (masterleinad) [inline cmake/KokkosConfig.cmake.in:23]:
```suggestion
    IF(NOT IS_NVCC AND NOT CMAKE_CXX_COMPILER_ID STREQUAL Clang AND
```

Comment (jrmadsen) [reply cmake/KokkosConfig.cmake.in]: Yea just caught that caught that

**Comment by masterleinad** [comment]
The architecture auto-detection fails (maybe unsurprisingly) with
```
/usr/bin/c++  -DSM_ONLY  -Werror    -std=c++11 -o CMakeFiles/cmTC_35a1f.dir/cuda_compute_capability.cc.o -c /tmp/ArborX/my_docker/kokkos/cmake/compile_tests/cuda_compute_capability.cc
/tmp/ArborX/my_docker/kokkos/cmake/compile_tests/cuda_compute_capability.cc: In function 'int main()':
/tmp/ArborX/my_docker/kokkos/cmake/compile_tests/cuda_compute_capability.cc:49:3: error: 'cudaDeviceProp' was not declared in this scope
   cudaDeviceProp device_properties;
   ^~~~~~~~~~~~~~
/tmp/ArborX/my_docker/kokkos/cmake/compile_tests/cuda_compute_capability.cc:50:9: error: 'cudaError_t' does not name a type; did you mean 'error_t'?
   const cudaError_t error = cudaGetDeviceProperties(&device_properties,
         ^~~~~~~~~~~
         error_t
/tmp/ArborX/my_docker/kokkos/cmake/compile_tests/cuda_compute_capability.cc:52:7: error: 'error' was not declared in this scope
   if (error != cudaSuccess) {
       ^~~~~
/tmp/ArborX/my_docker/kokkos/cmake/compile_tests/cuda_compute_capability.cc:52:7: note: suggested alternative: 'perror'
   if (error != cudaSuccess) {
       ^~~~~
       perror
/tmp/ArborX/my_docker/kokkos/cmake/compile_tests/cuda_compute_capability.cc:52:16: error: 'cudaSuccess' was not declared in this scope
   if (error != cudaSuccess) {
                ^~~~~~~~~~~
/tmp/ArborX/my_docker/kokkos/cmake/compile_tests/cuda_compute_capability.cc:53:36: error: 'cudaGetErrorString' was not declared in this scope
     std::cout << "CUDA error: " << cudaGetErrorString(error) << '\n';
                                    ^~~~~~~~~~~~~~~~~~
/tmp/ArborX/my_docker/kokkos/cmake/compile_tests/cuda_compute_capability.cc:57:7: error: 'device_properties' was not declared in this scope
       device_properties.major * 10 + device_properties.minor;
       ^~~~~~~~~~~~~~~~~
```

**Comment by jrmadsen** [comment]
Oh, well this isn't too complicated of a fix. 

```cmake
IF(KOKKOS_ENABLE_CUDA AND NOT CUDA_ARCH_ALREADY_SPECIFIED)
  TRY_RUN(
    _RESULT
    _COMPILE_RESULT
    ${_BINARY_TEST_DIR}
    ${CMAKE_CURRENT_SOURCE_DIR}/cmake/compile_tests/cuda_compute_capability.cc
    COMPILE_DEFINITIONS -DSM_ONLY
    RUN_OUTPUT_VARIABLE _CUDA_COMPUTE_CAPABILITY)
ENDIF()
```

I don't see a way to tell cmake to use the launcher so just need to either:

- add an if block around using `try_run` and then do it manually via `execute_process`
- `enable_language(CUDA)` + `CMAKE_TRY_COMPILE_TARGET_TYPE CUDA`

@masterleinad @jjwilke Thoughts? Preferences?

**Comment by masterleinad** [comment]
Currently, the auto-detection doesn't work with `clang` as a compiler either. It would be nice to come up with a solution that works for compilers different from `nvcc_wrapper` but that could be done separately and I don't think it's crucial for this pull request.

AFAICT, `CMAKE_TRY_COMPILE_TARGET_TYPE` can only differentiate between executables and libraries (https://cmake.org/cmake/help/latest/variable/CMAKE_TRY_COMPILE_TARGET_TYPE.html#variable:CMAKE_TRY_COMPILE_TARGET_TYPE) but can't be used to specify the language resp. the compiler to use. I guess the only way really is `execute_process`.

**Comment by jrmadsen** [comment]
@masterleinad Been too busy lately to figure out the auto-detection but I agree it would be nice to have in the future.

File: .jenkins
Comment (masterleinad) [inline .jenkins:200]:
```
cmake -DCMAKE_BUILD_TYPE=Release -DCMAKE_CXX_FLAGS=-Werror -DCMAKE_CXX_STANDARD=17 -DKokkos_ENABLE_COMPILER_WARNINGS=ON -DKokkos_ENABLE_OPENMP=ON -DKokkos_ENABLE_CUDA=ON -DKokkos_ENABLE_CUDA_LAMBDA=OFF -DKokkos_ENABLE_CUDA_UVM=ON -DKokkos_ENABLE_CUDA_RELOCATABLE_DEVICE_CODE=ON -DKokkos_ARCH_VOLTA70=ON -DCMAKE_INSTALL_PREFIX=/var/jenkins/workspace/Kokkos/build/../install ..
-- Setting default Kokkos CXX standard to 17
-- The CXX compiler identification is GNU 7.5.0
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ - works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Setting policy CMP0074 to use <Package>_ROOT variables
-- The project name is: Kokkos
-- Compiler Version: 11.0.167
-- kokkos_launch_compiler (/var/jenkins/workspace/Kokkos/bin/kokkos_launch_compiler) is enabled... Set Kokkos_LAUNCH_COMPILER_INFO=ON for more info.
-- Using -std=c++1z for C++17 standard as feature
CMake Error at cmake/kokkos_test_cxx_std.cmake:33 (MESSAGE):
  CMake wants to use -std=c++1z which is not supported by NVCC.  Using a more
  recent host compiler or a more recent CMake version might help.
Call Stack (most recent call first):
  cmake/kokkos_test_cxx_std.cmake:75 (kokkos_set_cxx_standard_feature)
  cmake/kokkos_tribits.cmake:201 (INCLUDE)
  CMakeLists.txt:161 (KOKKOS_SETUP_BUILD_ENVIRONMENT)
```
Apparently, `gcc-7.5.0` is still used for feature detection.

Comment (jrmadsen) [reply .jenkins]: Yea, just fixed that, it was setting `NVCC_WRAPPER_DEFAULT_COMPILER` instead of `CXX` in the environment

**Comment by masterleinad** [comment]
Retest this please.

**Comment by jrmadsen** [comment]
@masterleinad Hey just got back from vacation this week, is this ready for merging?

**Comment by masterleinad** [comment]
I think we should wait for this until after the release to have sufficient time to test it out.

File: cmake/kokkos_functions.cmake
Comment (jjwilke) [inline cmake/kokkos_functions.cmake:883]: We've discussed this before, but didn't bother with it. I think we're reaching a critical code duplication point with this function appearing in two places.

File: bin/kokkos_launch_compiler
Comment (jrmadsen) [reply bin/kokkos_launch_compiler:33]: The `shift` command drops the leading argument, thus `$2` became `$1`

File: bin/kokkos_launch_compiler
Comment (jrmadsen) [reply bin/kokkos_launch_compiler:45]: I sort of styled some of this after what I saw in nvcc_wrapper. It isn't intended to be an environment variable that anybody would really use, I just kind of put it there as a precaution in case it really needed to be set somewhere (like maybe some weird spack install or something)

File: cmake/KokkosConfig.cmake.in
Comment (jrmadsen) [reply cmake/KokkosConfig.cmake.in:22]: Typo

File: cmake/KokkosConfigCommon.cmake.in
Comment (jrmadsen) [reply cmake/KokkosConfigCommon.cmake.in:120]: If there is one, I missed it in the docs. It won't break Fortran or C, that's why I pass CMAKE_CXX_COMPILER -- when the compiler command gets invoked it looks like this : `kokkos_launch_compiler g++ g++ ...` so all I have to do is compare the two and if they are not equal (e.g. `kokkps_launch_compiler g++ gcc`) then the script just invokes the original command

File: cmake/KokkosConfigCommon.cmake.in
Comment (jrmadsen) [reply cmake/KokkosConfigCommon.cmake.in:121]: Same thing, the compiler is just passed to the script so it can discern which commands to forward to nvcc_wrapper

**Comment by jrmadsen** [comment]
@masterleinad I'm not sure I see the merit in delaying it. It doesn't change anything at all if CXX is set to nvcc_wrapper, it can be explicitly disabled, and merging it will help find edge cases much easier.

File: cmake/kokkos_functions.cmake
Comment (jrmadsen) [reply cmake/kokkos_functions.cmake:883]: Yea, I wasn't a huge fan of doing it but I didn't want to make the PR more complicated by moving a bunch of existing stuff around.

File: cmake/KokkosConfigCommon.cmake.in
Comment (jrmadsen) [reply cmake/KokkosConfigCommon.cmake.in:110]: I added the hints so it would default to the launcher in the installation but I didn't want to hardcode anything in the event installations got moved around. That's usually not much of a concern of course but ignoring it has basically traumatized me bc of all the issues it caused with PyPi installations in timemory (the Python package I use to invoke CMake from setup.py installs to a dummy dir and then reorganizes things the way Python likes it)

File: cmake/kokkos_functions.cmake
Comment (jjwilke) [reply cmake/kokkos_functions.cmake:883]: Fair enough. I'm fine if we merge this PR as-is and then we do a "deduplication" PR.

File: bin/kokkos_launch_compiler
Comment (jjwilke) [inline bin/kokkos_launch_compiler:33]: Ha, sorry, I meant add a comment here about the logic of checking the C++ compiler against itself and why you were doing it. I didn't actually care about the 'not not'.

Comment (jrmadsen) [reply bin/kokkos_launch_compiler:50]: Ah gotcha

File: cmake/KokkosConfigCommon.cmake.in
Comment (jrmadsen) [reply cmake/KokkosConfigCommon.cmake.in:110]: Resolved?

File: bin/kokkos_launch_compiler
Comment (jjwilke) [inline bin/kokkos_launch_compiler:58]: This prints a bit messy during configure when nvcc_wrapper isn't found. Maybe we can pipe `2> /dev/null`?

File: cmake/KokkosConfig.cmake.in
Comment (jjwilke) [inline cmake/KokkosConfig.cmake.in:25]: I think it probably makes sense to print a STATUS message here to let the user know what is going on.

File: cmake/KokkosConfig.cmake.in
Comment (jjwilke) [inline cmake/KokkosConfig.cmake.in:19]: I'm going back and forth on what the default mode should be. I'm wondering if it might be better to reverse this - make the person ask for an `NVCC_LAUNCHER` component to globally apply and make the default that no wrapper is added.

separable_compilation is a bit confusing on second read, although I can't think of a clearly better name at the moment.
separable_compilation doesn't seem very descriptive to me on second read.

**Review by jjwilke — DISMISSED** [review]
I'm happy with my tests. A bit of cleanup on error messages/status messages and I'm basically ready to approve.

Curious about thoughts on changing default for the launcher script to be globally applied.

**Comment by masterleinad** [comment]
Retest this please.

File: bin/kokkos_launch_compiler
Comment (jrmadsen) [reply bin/kokkos_launch_compiler:58]: That's strange... `which nvcc_wrapper` should just return nothing if it is not in the path and and that funky bash syntax just means don't set `BIN_NVCC_WRAPPER` if it is already set or if the string is empty.

Comment (jrmadsen) [reply bin/kokkos_launch_compiler:58]: I can't reproduce this issue. Did you do this: 
```console
: ${BIN_NVCC_WRAPPER:-"$(which nvcc_wrapper)"} 2> /dev/null
``` 
or this: 
```console
: ${BIN_NVCC_WRAPPER:-"$(which nvcc_wrapper 2> /dev/null)"}
```

Comment (jjwilke) [reply bin/kokkos_launch_compiler:58]: Yeah, I get an ugly block with my entire PATH and a 'no nvcc_wrapper found` error on some of our systems. Just arguing aesthetics, not function, at this point. But it would be confusing for users I think.

Comment (jjwilke) [reply bin/kokkos_launch_compiler:58]: The latter.

File: cmake/KokkosConfig.cmake.in
Comment (jrmadsen) [reply cmake/KokkosConfig.cmake.in:19]: IMO, I think the default should be the way it is because this makes kokkos + cuda and cmake act a lot more like a traditional project. With this setup, a novice to kokkos that doesn't know anything about nvcc_wrapper (or somebody that hasn't read the docs) and thus doesn't set the compiler at all, can just run `ccmake ... && make install` and then add a `find_package(Kokkos)` to their project in blissful ignorance of `nvcc_wrapper` bc it will (hopefully) _just work_. Any project that doesn't want to allow this, can just set a local variable in their CMake code before the find_package and their users won't be able to override it.

Comment (jrmadsen) [reply cmake/KokkosConfig.cmake.in:19]: I chose `separable_compilation` bc:

- We are effectively enabling files designated as the same language in CMake to be sent to separate compilers
- This option is specific to CUDA + NVCC which has a concept of separable compilation so the name will potentially be easy to associate and remember

File: cmake/KokkosConfig.cmake.in
Comment (jrmadsen) [reply cmake/KokkosConfig.cmake.in:26]: Done. Let me know if that works

File: bin/kokkos_launch_compiler
Comment (jrmadsen) [reply bin/kokkos_launch_compiler]: Went with the first

Comment (jjwilke) [reply bin/kokkos_launch_compiler]: Just pulled and looks like you did the latter? I could be losing my mind.

`: ${BIN_NVCC_WRAPPER:-"$(which nvcc_wrapper 2> /dev/null)"}`

Anyhow, output is suppressed now. Looks good.

Comment (jrmadsen) [reply bin/kokkos_launch_compiler]: I did the former, committed and then saw your comment and re-committed

**Comment by jjwilke** [comment]
Still thinking about this. My concern now is if 2 projects import Kokkos.  One of them asks for separable_compilation. The other does not.  You now won't get separable compilation because the global property has been set.

A counter-proposal would be this:
1) Keep everything you have, except that you can now ONLY point to nvcc_wrapper from your Kokkos installation. These are not allowed to be inconsistent.
2) Have Kokkos add a transitive flag like `--kokkos-dependence` to everything.  The compiler launcher can actually look for this. If this flag does not appear, the compiler launcher can actually just pass the args through.  If the flag does appear, use nvcc_wrapper.

This PR would be a better place to add the `--kokkos-dependence` flag. 

Now we have the separable compilation problem fixed. No one needs to ask for it.  If you depend on Kokkos, you go to the wrapper. If not, you're good to go. And you don't need to worry about two different Kokkos projects thrashing each other's global properties (of course, I could be misunderstanding global scope rules in find_package).

**Comment by jrmadsen** [comment]
@jjwilke I like it. What about using a compiler definition instead of a compiler flag? E.g. `-DKOKKOS_DEPENDENCE`. The benefit is that it wouldn't have to be removed from the args in the launcher and it could actually help downstream projects (in the 2 project scenario you mentioned) detect whether Kokkos is being used by parent project.

**Comment by jjwilke** [comment]
Yeah, yeah, yeah. That's better. I think this finally checks all the boxes.

**Comment by jrmadsen** [comment]
@jjwilke In light of these changes, it maybe possible to use the [target-based `RULE_LAUNCH_COMPILE](https://cmake.org/cmake/help/latest/prop_tgt/RULE_LAUNCH_COMPILE.html), query the global `RULE_LAUNCH_COMPILE`, and work that into the launcher supporting it, and thus support ccache (@dalg24). Probably should be done in another PR to avoid having to changing this PR too significantly but I think creating a `kokkos_launch_ccache` alternative for that situation might be the way to go.

**Comment by jrmadsen** [comment]
Also, in light of the changes, it appears the target-based `RULE_LAUNCH_COMPILE` and `RULE_LAUNCH_LINK` is the way to go instead of the global ones.

**Comment by berselius** [comment]
Dumb question from a cmake novice here - just want to make sure I'm parsing this correctly. After this PR, would the new use pattern for using kokkos in a single package of a larger code look like

base_level CMakeLists.txt
-----------
CMAKE_CXX_COMPILER=gcc
find_package(Kokkos COMPONENTS separable_compilation)

package_that_uses_kokkos CMakeLists.txt
--------------------------
kokkos_compilation(TARGET package_that_uses_kokkos)

and the rest of the packages wouldn't have any changes?

**Comment by jrmadsen** [comment]
## Behavior

| `CMAKE_CXX_COMPILER` | Compiler command | Notes |
| --------------------------- | -------------------- | --- |
| `nvcc_wrapper` | `nvcc_wrapper ...` | `kokkos_launch_compiler` is not used |
| `clang++` | `clang++ ...` | `kokkos_launch_compiler` is not used |
| `g++` | `kokkos_launch_compiler g++ ${EXE} ...` | When `${EXE} == g++` and `-DKOKKOS_DEPENDENCE` is found in `...`, redirects `...` to `nvcc_wrapper` |
|   |   | When `${EXE} != g++` or `-DKOKKOS_DEPENDENCE` is not found in `...`, execute `${CMD} ...` |
| `mpicxx` | `kokkos_launch_compiler mpicxx ${EXE} ...` | When `${EXE} == mpicxx` and `-DKOKKOS_DEPENDENCE` is found in `...`, redirects `...` to `nvcc_wrapper` |
|   |   | This sets `NVCC_HOST_COMPILER` so it should preserve previous behavior |

Note: `kokkos_launch_compiler` sets `NVCC_HOST_COMPILER` via `: ${NVCC_WRAPPER_DEFAULT_COMPILER:=${CXX_COMPILER}}` so setting `NVCC_WRAPPER_DEFAULT_COMPILER` in the environment _should_ preserve all previous behavior.

## Current Known Issues

In order for `kokkos_launch_compiler` to also be applied at the link stage, we have to add `KOKKOS_DEPENDENCE` as a definition in the link flags. This causes a minor issue because `target_compile_definitions(... PUBLIC KOKKOS_DEPENDENCE)` does not get added to the link flags. Thus we need to do `target_link_options(... PUBLIC -DKOKKOS_DEPENDENCE)`. The problem is the `-D` is not entirely portable, Windows, for example, wants `/D` instead of `-D` because microsoft just loves to make life difficult. Adding a specific case for `if(MSVC)` is easy but we could potentially have other issues down the line if some other compiler does not used `-D` for pre-processor definitions. Although, how are compiling Kokkos + CUDA on Windows? Is that even supported yet bc unless Windows sneakily added support for bash scripts, we aren't able to use nvcc_wrapper anyway.

**Comment by jrmadsen** [comment]
@berselius Yes. Although with the more recent changes, there is a `-DKOKKOS_DEPENDENCE` flag that is added to the Kokkos cmake target and when that is present, any target that depends on kokkos will redirect to nvcc_wrapper and if not, it won't. So the separable compilation isn't really necessary anymore

**Comment by crtrott** [comment]
ok we need to still figure out how to resolve the issue with current use cases (actually primary use case at Sandia) where we set OMPI_CXX=nvcc_wrapper. With this change we get a recursive dependency: nvcc_wrapper -> mpicxx -> nvcc_wrapper -> mpicxx ...

**Comment by crtrott** [comment]
Possible solution: don't inject launch compiler if the compiler can do CUDA on its own.

**Comment by crtrott** [comment]
The CUDA Arch detection doesn't work without a CUDA compiler. And it looks like downstream you have to set CMAKE_POLICY CMP0057 again?

**Comment by crtrott** [comment]
Retest this please.

File: bin/kokkos_launch_compiler
Comment (crtrott) [inline bin/kokkos_launch_compiler:68]: -e missing

**Review by dalg24 — APPROVED** [review]
I think this looks good

I moved the testing into the CUDA 11 build
https://github.com/kokkos/kokkos/blob/6bf9c87247af0f78fe2932e0d001c736708faab8/.jenkins#L184-L245
I purposefully use `nvcc_wrapper` directly in the "install testing" step.

**Pull request merged**
